### ISSUE TITLE
HWRP does not restart filebeat service

### ISSUE THREAD

**Issue by CVTJNII** [vkhatri/chef-filebeat#60]
When a change is flagged by the HWRP the filebeat service doesn't restart.  To verify this I recorded the pids before a change run:

```
vagrant@default-ubuntu-1404:~$ ps awwux | grep [b]eat
root      2055  0.0  0.0   9392     0 ?        S    Jun30   0:00 /usr/bin/filebeat-god -r / -n -p /var/run/filebeat.pid -- /usr/bin/filebeat -c /etc/filebeat/filebeat.yml
root      2062  0.0  1.5 261132  5884 ?        Sl   Jun30   0:19 /usr/bin/filebeat -c /etc/filebeat/filebeat.yml
```

Ran Chef changing action on the provider to :delete

```
-----> Starting Kitchen (v1.5.0)
-----> Converging <default-ubuntu-1404>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 4.2.1...
       Removing non-cookbook files before transfer
       Preparing validation.pem
       Preparing client.rb
-----> Chef Omnibus installation detected (install only if missing)
       Transferring files to <default-ubuntu-1404>
       Starting Chef Client, version 12.11.18
       resolving cookbooks for run list: ["apt", "filebeat::default"]
       Synchronizing Cookbooks:
         - apt (4.0.1)         
         - yum (3.11.0)        
         - ms_dotnet (2.6.2)   
         - chef_handler (1.4.0)
         - filebeat (0.3.5)    
         - windows (1.44.1)    
         - compat_resource (12.10.6)
         - powershell (3.3.2)  
       Installing Cookbook Gems:
       Compiling Cookbooks...  
       [2016-07-01T15:54:15+00:00] WARN: Chef::Provider::AptRepository already exists!  Cannot create deprecation class for LWRP provider apt_repository from cookbook apt
       [2016-07-01T15:54:15+00:00] WARN: AptRepository already exists!  Deprecation class overwrites Custom resource apt_repository from cookbook apt
       [2016-07-01T15:54:15+00:00] WARN: Please use the powershell_script resource in Chef Client 11.6+
       Converging 18 resources 
       Recipe: apt::default    
         * file[/var/lib/apt/periodic/update-success-stamp] action nothing (skipped due to action :nothing)
         * apt_update[periodic] action periodic (up to date)
         * execute[apt-get update] action nothing (skipped due to action :nothing)
         * execute[apt-get autoremove] action nothing (skipped due to action :nothing)
         * execute[apt-get autoclean] action nothing (skipped due to action :nothing)
         * directory[/var/cache/local] action create (up to date)
         * directory[/var/cache/local/preseeding] action create (up to date)
         * template[/etc/apt/apt.conf.d/10recommends] action create (up to date)
         * apt_package[apt-transport-https] action install (up to date)
       Recipe: filebeat::install_package
         * apt_repository[beats] action add
           * remote_file[/tmp/kitchen/cache/GPG-KEY-elasticsearch] action create (up to date)
           * execute[apt-key add /tmp/kitchen/cache/GPG-KEY-elasticsearch] action run (skipped due to not_if)
           * execute[apt-cache gencaches] action nothing (skipped due to action :nothing)
           * apt_update[beats] action nothing (skipped due to action :nothing)
           * file[/etc/apt/sources.list.d/beats.list] action create (up to date)
            (up to date)       
         * apt_package[filebeat] action install (up to date)
       Recipe: filebeat::config
         * directory[/etc/filebeat/conf.d] action create (up to date)
         * file[/etc/filebeat/filebeat.yml] action create (up to date)
         * file[prospector-system_logs] action create (up to date)
         * powershell[install filebeat as service] action run (skipped due to only_if)
         * ruby_block[delay filebeat service start] action run
           - execute the ruby block delay filebeat service start
         * service[filebeat] action enable (up to date)
         * service[filebeat] action nothing (skipped due to action :nothing)
       Recipe: filebeat::default
         * filebeat_prospector[messages] action delete
         Recipe: <Dynamically Defined Resource>
           * file[prospector_messages] action delete
             - delete file /etc/filebeat/conf.d/prospector-messages.yml

       Recipe: filebeat::config
         * service[filebeat] action start (up to date)

       Running handlers:       
       Running handlers complete
       Chef Client finished, 3/26 resources updated in 07 seconds
       Finished converging <default-ubuntu-1404> (0m18.37s).
-----> Kitchen is finished. (0m18.84s)
```

Note that the file was deleted correctly but the service action was :start and marked as up to date.  Per [provider_prospector.rb#L64](https://github.com/vkhatri/chef-filebeat/blob/master/libraries/provider_prospector.rb#L64) I would expect the action to be restart.

Check the pids again.  Note that they are the same signifying no restart took place:

```
vagrant@default-ubuntu-1404:~$ ps awwux | grep [b]eat
root      2055  0.0  0.0   9392     0 ?        S    Jun30   0:00 /usr/bin/filebeat-god -r / -n -p /var/run/filebeat.pid -- /usr/bin/filebeat -c /etc/filebeat/filebeat.yml
root      2062  0.0  1.5 261132  5884 ?        Sl   Jun30   0:19 /usr/bin/filebeat -c /etc/filebeat/filebeat.yml
```

As I test I manually restarted filebeat and the pids do change, it is not a HUP.

```
vagrant@default-ubuntu-1404:~$ ps awwux | grep [b]eat
root      2055  0.0  0.0   9392     0 ?        S    Jun30   0:00 /usr/bin/filebeat-god -r / -n -p /var/run/filebeat.pid -- /usr/bin/filebeat -c /etc/filebeat/filebeat.yml
root      2062  0.0  1.5 261132  5884 ?        Sl   Jun30   0:19 /usr/bin/filebeat -c /etc/filebeat/filebeat.yml
vagrant@default-ubuntu-1404:~$ sudo /etc/init.d/filebeat restart
 * Restarting Sends log files to Logstash or directly to Elasticsearch. filebeat
   ...done.
vagrant@default-ubuntu-1404:~$ ps awwux | grep [b]eat
root      4450  0.0  0.0   9392   180 pts/0    S    15:54   0:00 /usr/bin/filebeat-god -r / -n -p /var/run/filebeat.pid -- /usr/bin/filebeat -c /etc/filebeat/filebeat.yml
root      4451  1.5  2.7 294968 10364 pts/0    Sl   15:54   0:00 /usr/bin/filebeat -c /etc/filebeat/filebeat.yml
```

I expect this is a wider bug than just the HWRP as the code in the HWRP looks fine, but I've only confirmed it via the provider.

**Comment by CVTJNII** [comment]
No, it's the HWRP.  You cannot notify from the provider like this because it won't be added to the notification stack as it isn't run at compile time.  Dynamic resources cannot use notifies unfortunately.

The usual answer is to call resources(service: filebeat).run_action(:restart) if updated_by_last_action? is true, but that is immediate.  I don't believe it can be delayed.  (Does it need to be?)

EDIT: resources() isn't available in the provider scope.  I'm not sure how to do this without getting deep into Chef's guts, which will be brittle.

**Comment by CVTJNII** [comment]
Is there a reason this is a HWRP and not a definition?  A definition around the file provider seems like a simpler solution without as many edge cases.

EDIT: I believe #58 would be a non-issue if this was a definition as well.

**Comment by vkhatri** [comment]
@CVTJNII Resolved in v0.4.0. Sorry for the delay.
